### PR TITLE
[FIX] bus: add kill switch for outdated tabs

### PR DESCRIPTION
**Pull request by tsm-odoo**
The WebSocket worker is tied to a specific version. When this version
changes, the worker's URL updates, creating a new worker. The old
worker remains active as long as clients are connected to it. The
updates may take time to propagate, leading to both the old and new
workers being active simultaneously.

This PR introduces a mechanism to ensure that client code updates are
applied quicker.

When the worker opens a WebSocket connection (typically after the
server comes back online following an update), it provides the server
with its version. If the version is outdated, the server closes the
connection. Consequently, the worker will no longer open new WebSocket
connections.

Clients connected to an outdated worker will receive a notification
prompting them to reload the tab. Additionally, outdated tabs will not
be considered for the main tab election process.

### CODE REVIEW

File: addons/bus/static/src/services/bus_service.js [hunk 1]
```diff
@@ -60,21 +61,41 @@ export const busService = {
...
+                }
+                case "initialized": {
+                    isInitialized = true;
+                    connectionInitializedDeferred.resolve();
+                    break;
+                }
+                case "outdated": {
+                    multiTab.unregister();
+                    notification.add(_t("The page appears to be out of date."), {
+                        title: _t("Refresh"),
```
Comment (seb-odoo) [inline addons/bus/static/src/services/bus_service.js:82]: I'm not sure if that's clear enough as a title

Comment (seb-odoo) [inline addons/bus/static/src/services/bus_service.js:82]: export .pot

File: addons/bus/static/src/services/bus_service.js [hunk 2]
```diff
@@ -60,21 +61,41 @@ export const busService = {
...
+                    });
+                    break;
                 }
-            } else if (type === "initialized") {
-                isInitialized = true;
-                connectionInitializedDeferred.resolve();
-                return;
             }
-            if (type !== "notification") {
+            if (OUTGOING_EVENTS.has(type)) {
```
Comment (seb-odoo) [inline addons/bus/static/src/services/bus_service.js:98]: This was previously a default, why the change (in stable)?

File: addons/bus/static/src/services/bus_service.js [hunk 3]
```diff
@@ -98,7 +119,9 @@ export const busService = {
                 uid = false;
             }
             send("initialize_connection", {
-                websocketURL: `${params.serverURL.replace("http", "ws")}/websocket`,
+                websocketURL: `${params.serverURL.replace("http", "ws")}/websocket?version=${
+                    session.websocket_worker_version
```
Comment (seb-odoo) [inline addons/bus/static/src/services/bus_service.js:123]: Will the session be properly filled? All cases of livechat (embed)?

File: addons/bus/static/src/workers/websocket_worker.js [hunk 4]
```diff
@@ -30,11 +30,7 @@ export const WEBSOCKET_CLOSE_CODES = Object.freeze({
     BAD_GATEWAY: 1014,
     SESSION_EXPIRED: 4001,
     KEEP_ALIVE_TIMEOUT: 4002,
-    RECONNECTING: 4003,
```
Comment (seb-odoo) [inline addons/bus/static/src/workers/websocket_worker.js:33]: No remove in stable

File: addons/bus/static/src/workers/websocket_worker.js [hunk 5]
```diff
@@ -30,11 +30,7 @@ export const WEBSOCKET_CLOSE_CODES = Object.freeze({
...
 });
-// Should be incremented on every worker update in order to force
-// update of the worker in browser cache.
-export const WORKER_VERSION = "1.0.7";
```
Comment (seb-odoo) [inline addons/bus/static/src/workers/websocket_worker.js:37]: No remove in stable

File: addons/im_livechat/views/im_livechat_channel_templates.xml [hunk 6]
```diff
@@ -96,6 +96,7 @@
                     window.odoo = {};
                 }
                 odoo.__session_info__ = Object.assign(odoo.__session_info__ || {}, {
+                    websocket_worker_version: <t t-out="json.dumps(info['websocket_worker_version'])"/>,
```
Comment (seb-odoo) [inline addons/im_livechat/views/im_livechat_channel_templates.xml:99]: won't be applied until `-u`, it should be in `options` instead

File: addons/bus/static/src/services/bus_service.js (hunk 2 above)
Comment (tsm-odoo) [reply addons/bus/static/src/services/bus_service.js:103]: Not clear what is emitted or not, some events are followed by "return" others are filtered with the condition here. Another type is added in this PR, which should not be emitted so I just explicitly listen what should be emitted. Nothing changes from an outside perspective, only events emitted are "connect", "disconnect", "reconnecting", "reconnect".

File: addons/bus/static/src/services/bus_service.js (hunk 3 above)
Comment (tsm-odoo) [reply addons/bus/static/src/services/bus_service.js:128]: I checked for: frontend, backend, public page, support page, embed live chat. All are properly filled.

File: addons/bus/i18n/bus.pot [hunk 7]
```diff
@@ -51,6 +51,11 @@ msgstr ""
 msgid "Display Name"
 msgstr ""
 
+#. module: bus
+#: model:ir.model,name:bus.model_ir_http
+msgid "HTTP Routing"
```
Comment (seb-odoo) [inline addons/bus/i18n/bus.pot:56]: Where is that coming from? Should probably not be in bus.

File: addons/bus/i18n/bus.pot [hunk 8]
```diff
@@ -6,8 +6,8 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Odoo Server saas~17.2\n"
 "Report-Msgid-Bugs-To: \n"
-"POT-Creation-Date: 2024-03-26 15:05+0000\n"
-"PO-Revision-Date: 2024-03-26 15:05+0000\n"
+"POT-Creation-Date: 2024-07-30 11:04+0000\n"
+"PO-Revision-Date: 2024-07-30 11:04+0000\n"
```
Comment (seb-odoo) [inline addons/bus/i18n/bus.pot:10]: no need for updating dates

File: addons/bus/static/src/services/bus_service.js (hunk 2 above)
Comment (seb-odoo) [reply addons/bus/static/src/services/bus_service.js:103]: But if for some reason somebody made a custo which relied on unsupported types, we're going to annoy them no?

**Comment by tsm-odoo** [comment]
@robodoo r+

**Comment by tsm-odoo** [comment]
@robodoo r+

rebase :arrow_up:
